gtsam/hybrid/tests/testDCSAM.cpp

```diff
@@ -0,0 +1,685 @@
+
+using namespace gtsam;
+
+using symbol_shorthand::C;
+using symbol_shorthand::D;
+using symbol_shorthand::L;
+using symbol_shorthand::X;
+
+const double tol = 1e-7;
+
+/******************************************************************************/
+
+/*
+ * Test simple DiscretePriorFactor. We construct a DiscreteFactor factor
+ * graph with a single binary variable d1 and a single custom
+ * DiscretePriorFactor p(d1) specifying the distribution:
+ * p(d1 = 0) = 0.1, p(d1 = 1) = 0.9.
+ *
+ * After constructing the factor graph, we validate that the marginals are
+ * correct (i.e. they match the input p(d1)) and the most probable estimate
+ * (computed by maximizing p(d1) over possible values of d1) is 1.
+ */
+TEST(DCSAM, DiscretePriorFactor) {
+  // Make an empty discrete factor graph
+  DiscreteFactorGraph dfg;
+
+  // We'll make a variable with 2 possible assignments
+  const size_t cardinality = 2;
+  DiscreteKey dk(D(1), cardinality);
+  const std::vector<double> probs{0.1, 0.9};
+
+  // Make a discrete prior factor and add it to the graph
+  DecisionTreeFactor dpf(dk, probs);
+  dfg.push_back(dpf);
+
+  // Solve
+  DiscreteValues mostProbableEstimate = dfg.optimize();
+
+  // Get the most probable estimate
+  const size_t mpeD = mostProbableEstimate.at(dk.first);
+
+  // Get the marginals
+  DiscreteMarginals discreteMarginals(dfg);
+  Vector margProbs = discreteMarginals.marginalProbabilities(dk);
+
+  // Verify that each marginal probability is within `tol` of the true marginal
+  for (size_t i = 0; i < dk.second; i++) {
+    bool margWithinTol = (abs(margProbs[i] - probs[i]) < tol);
+    EXPECT(margWithinTol);
+  }
+
+  // Ensure that the most probable estimate is correct
+  EXPECT_LONGS_EQUAL(mpeD, 1);
+}
+
+namespace discrete_mixture_fixture {
+// We'll make a variable with 2 possible assignments
+DiscreteKey dk(D(1), 2);
+
+// Key for single continuous variable
+Key x1 = X(1);
+
+// Make a factor for non-null hypothesis
+const double loc = 0.0;
+const double sigma1 = 1.0;
+auto prior_noise1 = noiseModel::Isotropic::Sigma(1, sigma1);
+auto f1 = std::make_shared<PriorFactor<double>>(x1, loc, prior_noise1);
+
+// Make a factor for null hypothesis
+const double sigmaNullHypo = 8.0;
+auto prior_noiseNullHypo = noiseModel::Isotropic::Sigma(1, sigmaNullHypo);
+auto fNullHypo =
+    std::make_shared<PriorFactor<double>>(x1, loc, prior_noiseNullHypo);
+}  // namespace discrete_mixture_fixture
+
+/*
+ * Test discrete optimization using a simple mixture.
+ *
+ * Construct a single factor (for a single variable) consisting of a
+ * discrete-conditional mixture.
+ * Here we have a "null hypothesis" consisting of
+ * a Gaussian with large variance and an "alternative hypothesis"
+ * consisting of a Gaussian with smaller variance.
+ * After initializing the continuous variable far away from
+ * the ground-truth solution (x1 = 0), the discrete hypothesis selector
+ * will initially choose the null hypothesis.
+ */
+TEST(DCSAM, DiscreteMixture) {
```

Comment:
Again, technically not a `DCSAM` test and the motivation was really more for my own understanding. You can keep it or toss it, but you certainly have my blessing to remove it if you want 😂